core/src/mindustry/input/Placement.java

```diff
@@ -54,6 +55,20 @@
         return points;
     }
 
+    public static Seq<Point2> upgradeLine(int startX, int startY, int endX, int endY){
+        Pools.freeAll(points);
+        points.clear();
+        Building building = world.tile(startX, startY).build;
+        points.add(Pools.obtain(Point2.class, Point2::new).set(startX, startY));
+        while((building.tile.x != endX || building.tile.y != endY)){
```

Comment:
why place double brackets here?

```suggestion
        while(building.tile.x != endX || building.tile.y != endY){
```